tasmota/xsns_73_hp303b.ino

```diff
@@ -29,104 +29,113 @@
 /*********************************************************************************************/
 
-bool HP303B_Read()
+bool HP303B_Read(uint8_t hp303b_idx)
 {
-  if (bhp303b_sensor.valid) { bhp303b_sensor.valid--; }
+  if (hp303b_sensor[hp303b_idx].valid) { hp303b_sensor[hp303b_idx].valid--; }
 
   float t;
   float p;
   int16_t ret;
 
-  ret = HP303BSensor.measureTempOnce(t, bhp303b_sensor.oversampling);
+  ret = HP303BSensor.measureTempOnce(t, hp303b_cfg.oversampling);
   if (ret != 0)
     return false;
 
-  ret = HP303BSensor.measurePressureOnce(p, bhp303b_sensor.oversampling);
+  ret = HP303BSensor.measurePressureOnce(p, hp303b_cfg.oversampling);
   if (ret != 0)
     return false;
 
-  bhp303b_sensor.temperature = (float)ConvertTemp(t);
-  bhp303b_sensor.pressure = (float)ConvertPressure(p) / 100; //conversion to hPa
+  hp303b_sensor[hp303b_idx].temperature = (float)ConvertTemp(t);
+  hp303b_sensor[hp303b_idx].pressure = (float)ConvertPressure(p) / 100; //conversion to hPa
 
-  bhp303b_sensor.valid = SENSOR_MAX_MISS;
+  hp303b_sensor[hp303b_idx].valid = SENSOR_MAX_MISS;
   return true;
 }
 
 /********************************************************************************************/
 
 void HP303B_Detect(void)
 {
-  for (uint32_t i = 0; i < sizeof(bhp303b_addresses); i++)
+  for (uint32_t i = 0; i < HP303B_MAX_SENSORS; i++)
   {
-    if (I2cActive(bhp303b_addresses[i])) { return; }
-
-    bhp303b_sensor.address = bhp303b_addresses[i];
+    if (I2cActive(HP303B_START_ADDRESS + i)) { return; }
 
-    HP303BSensor.begin( bhp303b_sensor.address);
-
-    if (HP303B_Read())
+    if (HP303BSensor.begin(HP303B_START_ADDRESS + i))
     {
-      I2cSetActiveFound(bhp303b_sensor.address, bhp303b_sensor.name);
-      bhp303b_sensor.type = 1;
+      hp303b_sensor[hp303b_cfg.count].address = HP303B_START_ADDRESS + i;
+      I2cSetActiveFound(hp303b_sensor[hp303b_cfg.count].address, hp303b_cfg.types);
+      hp303b_cfg.count++;
       break;
     }
   }
 }
 
 void HP303B_EverySecond(void)
 {
-  if (uptime &1) {
-    if (!HP303B_Read()) {
-      AddLogMissed(bhp303b_sensor.name, bhp303b_sensor.valid);
+  for (uint32_t i = 0; i < hp303b_cfg.count; i++) {
+    if (uptime &1) {
+    if (!HP303B_Read(i)) {
+      AddLogMissed(hp303b_cfg.types, hp303b_sensor[i].valid);
     }
   }
+  }
 }
 
 void HP303B_Show(bool json)
 {
-  if (bhp303b_sensor.valid)
-  {
-    char str_temperature[33];
-    dtostrfd(bhp303b_sensor.temperature, Settings.flag2.temperature_resolution, str_temperature);
-    char str_pressure[33];
-    dtostrfd(bhp303b_sensor.pressure, Settings.flag2.pressure_resolution, str_pressure);
+  for (uint32_t i = 0; i < hp303b_cfg.count; i++) {
+    char sensor_name[12];
+    strlcpy(sensor_name, hp303b_cfg.types, sizeof(sensor_name));
+    if (hp303b_cfg.count > 1) {
+      snprintf_P(sensor_name, sizeof(sensor_name), PSTR("%s%c0x%02X"), sensor_name, IndexSeparator(), hp303b_sensor[i].address); // MCP9808-18, MCP9808-1A  etc.
+    }
```

Comment:
check all the comments. Some are from MCP9808

Reply:
Fixed